drivers/mfrc522/mfrc522.c

```diff
@@ -0,0 +1,2278 @@
+ * @param[in] data        Data to transfer in step 2
+ *
+ * @return MFRC522_STATUS_OK on success, MFRC522_STATUS_??? otherwise
+ * @return MFRC522_STATUS_INVALID if command was not out of
+ *         [MFRC522_PICC_CMD_MF_INCREMENT, MFRC522_PICC_CMD_MF_DECREMENT,
+ *         MFRC522_PICC_CMD_MF_RESTORE], MFRC522_STATUS_??? otherwise
+ */
+static mfrc522_status_code_t
+mfrc522_mifare_two_step_helper(mfrc522_t *dev, mfrc522_picc_command_t command,
+                               uint8_t block_addr, int32_t data);
+
+void mfrc522_pcd_set_register_bitmask(mfrc522_t *dev,
+                                      mfrc522_pcd_register_t reg, uint8_t mask)
+{
+    uint8_t tmp;
+    mfrc522_device_read(dev, reg, &tmp);
+
+    /* set bit mask */
+    mfrc522_device_write(dev, reg, tmp | mask);
+}
+
+void mfrc522_pcd_clear_register_bitmask(mfrc522_t *dev,
+                                        mfrc522_pcd_register_t reg, uint8_t mask)
+{
+    uint8_t tmp;
+    mfrc522_device_read(dev, reg, &tmp);
+
+    /* clear bit mask */
+    mfrc522_device_write(dev, reg, tmp & (~mask));
+}
+
+mfrc522_status_code_t mfrc522_pcd_calculate_crc(mfrc522_t *dev,
+                                                uint8_t *data,
+                                                uint8_t length,
+                                                uint8_t *result)
+{
+    /* Stop any active command */
+    mfrc522_device_write(dev, MFRC522_REG_COMMAND, MFRC522_CMD_IDLE);
+
+    /* Clear the CRCIRq interrupt request bit */
+    mfrc522_device_write(dev, MFRC522_REG_DIV_IRQ, MFRC522_BIT_DIV_IRQ_CRC_IRQ);
+
+    /* FlushBuffer = 1, FIFO initialization */
+    mfrc522_device_write(dev, MFRC522_REG_FIFO_LEVEL, 0x80);
+
+    /* Write data to the FIFO */
+    mfrc522_device_write_n(dev, MFRC522_REG_FIFO_DATA, length, data);
+
+    /* Start the calculation */
+    mfrc522_device_write(dev, MFRC522_REG_COMMAND, MFRC522_CMD_CALC_CRC);
+
+    /* Wait for the CRC calculation to complete */
+    for (uint16_t i = 5000; i > 0; i--) {
+
+        /* 5000 * 18 us sums up to 90 ms */
+        ztimer_sleep(ZTIMER_USEC, 18);
+
+        uint8_t n;
+        mfrc522_device_read(dev, MFRC522_REG_DIV_IRQ, &n);
+
+        /* CRCIRq bit set - calculation done */
+        if (n & MFRC522_BIT_DIV_IRQ_CRC_IRQ) {
+            /* Stop calculating CRC for new content in the FIFO */
+            mfrc522_device_write(dev, MFRC522_REG_COMMAND, MFRC522_CMD_IDLE);
+            /* Transfer the result from the registers to the result buffer */
+            mfrc522_device_read(dev, MFRC522_REG_CRC_RESULT_LSB, &result[0]);
+            mfrc522_device_read(dev, MFRC522_REG_CRC_RESULT_MSB, &result[1]);
+
+            return MFRC522_STATUS_OK;
+        }
+    }
+
+    /* 90 ms passed and nothing happened. Communication with the MFRC522 might be down. */
+    return MFRC522_STATUS_TIMEOUT;
+}
+
+int mfrc522_pcd_init(mfrc522_t *dev, const mfrc522_params_t *params)
+{
+    assert(dev);
```

Comment:
Probably would be good to perform this check on the other functions as well.

Reply:
I added assertions to all functions with `mfrc522_t *dev` parameter